File: 0930_Binary Subarrays With Sum/BinarySubarraysWithSum.java
class Solution {
    public int numSubarraysWithSum(int[] nums, int goal) {
        Map<Integer,Integer> map = new HashMap<>();
        map.put(0,1);
        int presum = 0;
        int ans = 0;
        for(int v : nums) {
            presum += v;
            ans += map.getOrDefault(presum - goal, 0);
            map.put(presum, map.getOrDefault(presum, 0) + 1);
        }
        
        return ans;
    }
}

File: 0930_Binary Subarrays With Sum/BinarySubarraysWithSum.cs
public class Solution {
    public int NumSubarraysWithSum(int[] nums, int goal) {
        //presum, occurrence 
        var dict = new Dictionary<int,int>();
        dict.Add(0,1);
        var presum = 0;
        var ans = 0;
        for(int i=0;i<nums.Length;i++)
        {
            presum += nums[i];
            if(dict.ContainsKey(presum - goal)) ans += dict[presum - goal];
            
            if(!dict.ContainsKey(presum)) dict.Add(presum, 0);
            dict[presum]++;
            
            
        }
        
        return ans;
    }
}